sympy/physics/control/tests/test_lti.py

```diff
@@ -577,53 +574,186 @@ def test_Feedback_construction():
         (a2*s + p)*(s**2 + 2*s*wn*zeta + wn**2))*(s**2 + 2*s*wn*zeta + wn**2), s)
-    assert Feedback(tf1*tf2, tf5).doit() == \
+    assert Feedback(TF1*TF2, tf5).doit() == \
         TransferFunction(k*(a0 + s)*(s**2 + 2*s*wn*zeta + wn**2), (k*(-a0 + a1*s**2 + a2*s) + \
         (a0 + s)*(s**2 + 2*s*wn*zeta + wn**2))*(s**2 + 2*s*wn*zeta + wn**2), s)
     assert Feedback(tf4, tf6).doit() == \
         TransferFunction(p*(p + s)*(a0*p + p**a1 - s), p*(p*(p + s) + (-p + s)*(a0*p + p**a1 - s)), p)
     assert -Feedback(tf4*tf6, TransferFunction(1, 1, p)).doit() == \
         TransferFunction(-p*(-p + s)*(p + s)*(a0*p + p**a1 - s), p*(p + s)*(p*(p + s) + (-p + s)*(a0*p + p**a1 - s)), p)
 
-    assert Feedback(tf1, tf2*tf5).rewrite(TransferFunction) == \
+    assert Feedback(TF1, TF2*tf5).rewrite(TransferFunction) == \
         TransferFunction((a0 + s)*(s**2 + 2*s*wn*zeta + wn**2), (k*(-a0 + a1*s**2 + a2*s) + \
         (a0 + s)*(s**2 + 2*s*wn*zeta + wn**2))*(s**2 + 2*s*wn*zeta + wn**2), s)
     assert Feedback(TransferFunction(1, 1, p), tf4).rewrite(TransferFunction) == \
         TransferFunction(p, a0*p + p + p**a1 - s, p)
+
+
+def test_TransferFunctionMatrix_construction():
+    tf5 = TransferFunction(a1*s**2 + a2*s - a0, s + a0, s)
+    tf4 = TransferFunction(a0*p + p**a1 - s, p, p)
+    tf6 = TransferFunction(a0*s**3 + a1*s**2 - a2*s, b0*p**4 + b1*p**3 - b2*s*p, s)
+    tf7 = TransferFunction(p, a0, p)
+
+    tfm1 = TransferFunctionMatrix([TF1, TF2], (2, 1), s)
+    assert tfm1.shape == (tfm1.outputs, tfm1.inputs) == (2, 1)
+    assert tfm1.args == ([TF1, TF2], (2, 1), s)
+    assert tfm1.var == s
+
+    tfm2 = TransferFunctionMatrix([-TF1, TF2], (2, 1), s)
+    assert tfm2.shape == (tfm2.outputs, tfm2.inputs) == (2, 1)
+    assert tfm2.args == ([-TF1, TF2], (2, 1), s)
+    assert tfm2.var == s
+
+    tfm3 = TransferFunctionMatrix([tf7], (1, 1), p)
+    assert tfm3.shape == (tfm3.outputs, tfm3.inputs) == (1, 1)
+    assert tfm3.var == p
+    assert tfm3.args == ([tf7], (1, 1), p)
+
+    tfm4 = TransferFunctionMatrix([TF3, tf5, tf6], (3, 1), s)
+    assert tfm4.shape == (tfm4.outputs, tfm4.inputs) == (3, 1)
+    assert tfm4.args == ([TF3, tf5, tf6], (3, 1), s)
+    assert tfm4.var == s
+
+    tfm5 = TransferFunctionMatrix([[TF1, -TF2], [TF3, tf5]], (2, 2), s)
+    assert tfm5.shape == (tfm5.outputs, tfm5.inputs) == (2, 2)
+    assert tfm5.args == ([[TF1, -TF2], [TF3, tf5]], (2, 2), s)
+
+    tfm6 = TransferFunctionMatrix([[TF1, TF2, TF3], [tf5, tf6, -tf6]], (2, 3), s)
+    assert tfm6.shape == (tfm6.outputs, tfm6.inputs) == (2, 3)
+    assert tfm6.args == ([[TF1, TF2, TF3], [tf5, tf6, -tf6]], (2, 3), s)
+
+    tfm7 = TransferFunctionMatrix([[TF1, TF2], [TF3, -tf5], [-tf5, TF2]], (3, 2), s)
+    assert tfm7.shape == (tfm7.outputs, tfm7.inputs) == (3, 2)
+    assert tfm7.args == ([[TF1, TF2], [TF3, -tf5], [-tf5, TF2]], (3, 2), s)
+
+    raises(TypeError, lambda: TransferFunction(Matrix([1, 2, 3]), (1, 1), s))
```

Comment:
Why is this invalid? Because of "Matrix"? I think you should follow duck typing principles here, just like other matrix-like classes. Any container class that looks like a list or list of lists with proper dimensions should work.

This example also seems to have two errors. The second one being the shape mismatch. These tests should expose one error at a time.

Reply:
> Any container class that looks like a list or list of lists with proper dimensions should work.

For now, I know how to allow `MutableMatrix` and `ImmutableMatrix` in the first argument. I'm not sure how any container class that looks like a list or list of lists will work.

Reply:
You should be able to do, at least, all of these:

```python
TransferFunctionMatrix([[tf1, tf2], [tf3, tf4]])
TransferFunctionMatrix(((tf1, tf2), (tf3, tf4)))
TransferFunctionMatrix(Matrix[[tf1, tf2], [tf3, tf4]])
```

Reply:
> You should be able to do, at least, all of these:
> 
> ```python
> TransferFunctionMatrix([[tf1, tf2], [tf3, tf4]])
> TransferFunctionMatrix(((tf1, tf2), (tf3, tf4)))
> TransferFunctionMatrix(Matrix[[tf1, tf2], [tf3, tf4]])
> ```

Allowing multiple types like this seems like poor API to me. I personally think that it is better just to require the argument to be a Matrix if it is supposed to be a Matrix.

If you do want to allow all of these types of input then the simplest way is just to call `Matrix` on the arguments.

I plan to deprecate using non-Expr classes in `Matrix`. They make it very difficult to improve matrix performance e.g. as in #19686 and lead to many bugs. Many methods need to be changed or disabled to make it possible to have non-Expr in a Matrix in a way that is vaguely consistent (see #19567 that tries to make a Matrix subclass with `Poly` as the elements).

So I think that `Matrix([[tf1, tf2], [tf3, tf4]])` should be an error. It is conceivable that we could create a mechanism that makes it possible for that to automatically translate to a different class such as `TransferFunctionMatrix` but that seems out of scope here.

My suggestion is to document that the argument should be a list of lists but allow any sequence of sequences by doing something like:
```
def __new__(self, rows):
    rows = [list(row) for row in rows]
    ...
```
(Or otherwise convert to tuple of tuples in a similar way)

I don't think that `Matrix([[tf1, tf2], [tf3, tf4]])` should be allowed because `TransferFunction` is non-Expr. It would make sense to have a method for converting a `TransferFunctionMatrix` to a `Matrix` but the elements of the resulting `Matrix` should be `Expr`. It could also make sense to have a function for converting from a `Matrix` to a `TransferFunctionMatrix` but it should be analogous to the way that a `TransferFunction` is constructed from an `Expr` (if there is already a way to do that).

Reply:
> Allowing multiple types like this seems like poor API to me. I personally think that it is better just to require the argument to be a Matrix if it is supposed to be a Matrix.

This is the API of NumPy and SymPy Matrix for over a decade, over 15 years for NumPy. Users of these libraries are accustom to this and I think we should support it here because of that. We have no performance issues at the moment, so you may be trying to solve a problem that doesn't exist.

Reply:
My point is this should be duck typed. If you want to deprecate what arguments are allowed in Matrix, that is a separate issue. TransferFunctionMatrix should accept anything that behaves like a lists of lists, in the iterable sense.

Reply:
> We have no performance issues at the moment

There are massive performance issues with matrices. Many of the performance issues listed on github were opened by you!

Here's a simple example:
```julia
In [1]: rref = lambda n: (randMatrix(n) + I*randMatrix(n)).rref()                                                                              

In [2]: for n in range(10): 
   ...:     %time ok = rref(n) 
   ...:                                                                                                                                        
CPU times: user 1.05 ms, sys: 770 µs, total: 1.82 ms
Wall time: 1.21 ms
CPU times: user 43.5 ms, sys: 6.59 ms, total: 50.1 ms
Wall time: 48.6 ms
CPU times: user 321 ms, sys: 11.3 ms, total: 332 ms
Wall time: 333 ms
CPU times: user 262 ms, sys: 1.24 ms, total: 263 ms
Wall time: 264 ms
CPU times: user 917 ms, sys: 3.87 ms, total: 921 ms
Wall time: 924 ms
CPU times: user 2.73 s, sys: 9.24 ms, total: 2.74 s
Wall time: 2.74 s
^C---------------------------------------------------------------------------
KeyboardInterrupt
```

In any case having non-Expr in a Matrix is buggy so I think that the solution here is not allowing `Matrix([[tf1, ...]])` If `tf1` is a non-Expr `TransferFunction`. We're quite close to being able to eliminate the other cases that have non-Expr in a Matrix around the sympy codebase after #19882

Reply:
> TransferFunctionMatrix should accept anything that behaves like a lists of lists, in the iterable sense.

Agreed. (That was my suggestion)

Reply:
Sounds like we were on the same page. We just shouldn't add unit tests for the Matrix() input then, so we don't encode that use.

Reply:
> There are massive performance issues with matrices. Many of the performance issues listed on github were opened by you!

I meant that we have no performance issues with `TransferFunctionMatrix`.

Reply:
> I meant that we have no performance issues with TransferFunctionMatrix.

Ah, yes. Well not yet.

I don't know what sort of calculations you intend to do with the transfer function matrix later but matrices of rational functions are handled *much* faster by #19686. See e.g. discussion in #19887.

The performance issues I refer to are actually performance issues in Matrix itself. Some of the obvious fixes are hard to implement because the implementation is limited by the fact that the class is also used with non-Expr elements in various parts of the codebase. Since it happens around sympy's own codebase I assume that a lot of downstream libraries/users are putting non-Expr into matrices as well so it's a significant backward compatibility problem. Possibly the solution is a new Matrix class (which also presents an opportunity to clean up a lot of other stuff).

Reply:
> I don't know what sort of calculations you intend to do with the transfer function matrix later but matrices of rational functions are handled much faster by #19686. See e.g. discussion in #19887.

I don't fully either. This is the problem of trying to design a class without drawing up a broad set of unit tests and examples to show what we'd like the class to do.

I'm a strong proponent of "documentation drive development" where we have some rich examples of intended use to be the guide. This PR is shooting in the dark.

Reply:
> This PR is shooting in the dark

If you think of it this way, can we just focus on writing a tutorial showing how to solve basic control problems using the already implemented functionality and also improving the documentation further in a different PR? A good set of examples can be added showing the usage of SISO transfer function and `Feedback` class. If some time would be left, I can continue my work on this PR.

Reply:
For the final weeks of GSoC I think that you should 1) finish this PR, 2) improve the documentation for the control module, and 3) add new methods to the classes you created to make them more useful (in that order of importance). But, you should be creating 2) documentation and examples as the first step to figuring out 1), otherwise noone knows what functionality you are trying to attain (including you!).